CI: Add ``ansys/actions/check-actions-security`` action and related fixes

ecoussoux-ansys opened the pull request:
This PR introduces the ``ansys/actions/check-actions-security`` action in the workflow file ``.github/workflows/ci_cd.yml`` and consequently in the CI of the ``magnet-segmentation-toolkit``.
This action is using ``zizmor`` to perform an audit of the workflows defined in the ``.github/workflows`` folder.

The PR addresses the findings surfaced by the ``zizmor`` audit on the workflow files (performed locally), resulting in the following changes:
* The argument ``persist-credentials: false`` is now systematically used with ``actions/checkout``,
* ``permissions`` are now defined on a job by job basis and none are granted at the workflow level. Generally, jobs that do not use secrets are not granted any specific permissions. In some cases, a comment is provided, describing why a permission is granted: Starting from ``v10.2.0``, the ``ansys/actions/check-actions-security`` is using ``v1.16.0`` of ``zizmor`` which is raising ``[undocumented-permissions]`` warnings if some permissions comments are not present (see https://github.com/ansys/actions/pull/1016 for example permission comments),
* Template expansions (``${{ ... }}``) are removed from plain run steps inside jobs. Inputs and relevant context variables are expanded in the ``env`` block instead,
* An anonymous definition for a job in the ``label.yml`` file is fixed,
* All the actions used in the workflow files are updated to their latest releases (``v10.2.3`` for ``ansys/actions``). All actions - both ``ansys/actions`` as well as external ones - had already been pinned with a commit SHA in #288,
* A potentially spoofable bot condition detected in ``ci_cd.yml`` (in the ``block-dependabot`` job) is ignored with an in-line comment. This approach to ignoring this one-off ``zizmor`` warning is preferred over introducing a ``zizmor.yml`` file as discussed in a review comment below, while the motivation for disregarding the warning is elaborated on in this PR: https://github.com/ansys/pyaedt/pull/6787.
*  In order to pass the dependabot-related checks performed by ``zizmor`` in the ``ansys/actions/check-actions-security`` from ``v10.2.0`` on, a ``[dependabot-cooldown]`` warning for the ``github-actions`` package-ecosystem is ignored with an in-line comment, as the cooldown feature is not available for the actions yet,
* In the ``tests_linux`` job defined in ``ci_cd.yml``, a command in the step ``AEDT Test`` is removed.

The method for introducing the ``ansys/actions/check-actions-security`` action is explained in details [here](https://actions.docs.ansys.com/version/stable/vulnerability-actions/#check-actions-security-action) and instructions for fixing common workflow vulnerabilities are provided [here](https://dev.docs.pyansys.com/how-to/vulnerabilities.html#addressing-common-vulnerabilities-in-github-actions).
The position of the action within the workflow has been defined based on the implementation of the ``check-vulnerabilities`` action in the ``toolkits-common`` project (see in particular https://github.com/ansys/pyaedt-toolkits-common/pull/313).

ecoussoux-ansys committed:
ci: Use argument persist-credentials: false with actions/checkout

ecoussoux-ansys committed:
ci: Fix anonymous-definition of commenter job in label.yml

ecoussoux-ansys committed:
ci: Avoid overly broad permissions at workflow level - Add specific ones for those using secrets

ecoussoux-ansys committed:
ci: Remove template expansions from inside job runs

ecoussoux-ansys committed:
ci: Introduce zizmor.yml file to ignore potentially spoofable bot condition detected in ci_cd.yml

ecoussoux-ansys committed:
ci: Add actions security check to workflow

ecoussoux-ansys committed:
ci: Update ansys/actions to v10.2.2

pyansys-ci-bot committed:
chore: adding changelog file 289.maintenance.md [dependabot-skip]

ecoussoux-ansys commented:
@SMoraisAnsys, I am not entirely sure what is the best way to deal with the detected "potentially spoofable bot condition" in the ``block-dependabot`` job.

First, I believe we want to ignore it for the time being, is that correct?

For now, I just reproduced the approach you implemented in https://github.com/ansys/pyaedt/pull/6787 in the same context. But since it's a single finding, could we not instead simply add a comment on the affected line in ``ci_cd.yml`` like ``# zizmor: ignore[bot-conditions]`` as described in the dev-guide [here](https://dev.docs.pyansys.com/how-to/vulnerabilities.html#ignoring-zizmor-findings) ? 
My main concern with the current approach is that if updates are made to ``ci_cd.yml``, the line where the error occurs could change, and it would then be easy to forget to update the ``zizmor`` config accordingly. Also if we use a simple comment, then we don't need the ``zizmor.yml`` at all, right?

ecoussoux-ansys committed:
ci: Fix code style errors and update zizmor.yml

ecoussoux-ansys commented:
I am putting this PR back to draft as it appears that I have a few more changes to perform in order to fix all the ``zizmor`` findings and have the ``ansys/actions/check-actions-security`` pass in the CI (extra findings of the type ``[undocumented-permissions]`` and ``[dependabot-cooldown]`` raised by the ``v1.16.0`` of ``zizmor`` used by the action starting from ``v10.2.0`` that I had overlooked locally).

ecoussoux-ansys committed:
ci: Use dependabot cooldown for github actions

ecoussoux-ansys committed:
ci: Provide permissions comments as requested by zizmor in undocumented-permissions findings

ecoussoux-ansys commented:
The last 2 commits have fixed the remaining ``zizmor`` findings, therefore putting the PR in ready for review again.
The PR description is updated to describe the additional changes performed.

Now the CI is still failing just as it did in #288 though.

SMoraisAnsys reviewed (changes requested):
Thanks for the changes @ecoussoux-ansys 
You are right about the zizmor comment, we can remove the zizmor YAML file and add the required comment in the ci_cd.yml file.
When I looked for a skip online I was led to https://docs.zizmor.sh/configuration/#settings but I think you approach is better in this case.
Can you also perform the same changes in pyaedt / other repos ?

ecoussoux-ansys committed:
ci: Ignore potentially spoofable bot condition in ci_cd.yml with inline comment and remove zizmor.yml file

ecoussoux-ansys commented:
> Thanks for the changes @ecoussoux-ansys You are right about the zizmor comment, we can remove the zizmor YAML file and add the required comment in the ci_cd.yml file. When I looked for a skip online I was led to https://docs.zizmor.sh/configuration/#settings but I think you approach is better in this case. Can you also perform the same changes in pyaedt / other repos ?

Thanks for the feedback @SMoraisAnsys !

I added the required comment in ``ci_cd.yml`` and removed the ``zizmor.yml`` file in 656f757. I also updated the PR description to reflect this.

As for applying this in the other projects:
- I adopted this approach directly when introducing ``ansys/actions/check-actions-security`` in the antenna toolkit (https://github.com/ansys/pyaedt-toolkits-antenna/pull/356) and will do the same when adding the action in the toolkit common,
- I will update pyaedt accordingly. 
- Regarding the radar toolkit: The implementation I introduced to the ``block-dependabot`` job for pyaedt in https://github.com/ansys/pyaedt/pull/6743  and which you reverted in https://github.com/ansys/pyaedt/pull/6787 was applied in https://github.com/ansys-internal/ansys-aedt-toolkits-radar-explorer/pull/211. It would probably make sense to update the job and deal with the ``zizmor`` warning as discussed here, but is it the right time to do so? I don't want to mess with the OSS approval for something so minor..
- pyaedt-examples and pyedb do not implement the ``block-dependabot`` job in their ``ci_cd.yml`` file, so no change will be needed there.

SMoraisAnsys commented on .github/workflows/ci_cd.yml:
```diff
@@ -49,7 +63,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
-      - name: Exit if dependabot triggered the workflow
+      - name: Exit if dependabot triggered the workflow # zizmor: ignore[bot-conditions] : Standard fix prevents any PR from dependabot to be possibly merged.
```
```suggestion
      - name: Exit if dependabot triggered the workflow # zizmor: ignore[bot-conditions] safe ignore because we exit if dependabot is the triggering action
```

ecoussoux-ansys replied:
That's a better explanation, thanks for the suggestion!

SMoraisAnsys commented on .github/workflows/ci_cd.yml:
```diff
@@ -369,9 +397,11 @@ jobs:
         timeout-minutes: 10
         run: |
-          export LD_LIBRARY_PATH=${{ env.ANSYSEM_ROOT251 }}/common/mono/Linux64/lib64:$LD_LIBRARY_PATH
+          export LD_LIBRARY_PATH=${ANSYSEM}/common/mono/Linux64/lib64:$LD_LIBRARY_PATH
```
I think that this is no longer required but we might have to double check. In pyaedt we have dropped them and it is now loaded when required through pyedb (which is the reason why we needed to add this library path)

ecoussoux-ansys replied:
Alright, thanks for the explanations.
In 41b31b6 I removed that command as well as the definitions of ``ANSYSEM`` and ``env.ANSYSEM_ROOT251`` in the ``tests_linux`` job.

SMoraisAnsys commented on .github/dependabot.yml:
```diff
@@ -33,6 +33,10 @@ updates:
+    cooldown:
+      default-days: 7
+      include:
+        - "*"  # Include all dependencies in cooldown
```
```suggestion
```
Unfortunately the cooldown is not (yet ?) available for github actions, see https://docs.github.com/en/code-security/dependabot/working-with-dependabot/dependabot-options-reference#configuration-of-cooldown
This might also need to be updated in payedt :/

ecoussoux-ansys replied:
I see, thanks for the hint.
In 41b31b6 I reverted this part and added an in-line comment to ignore the related ``[dependabot-cooldown]`` warning from ``zizmor`` on the github actions.
Noted to do the same in pyaedt!

SMoraisAnsys replied:
Following the dependabot PR this weekend I added the missing comment to ignore the warning on dependabot. Thanks @ecoussoux-ansys

SMoraisAnsys reviewed (changes requested):
Thanks for the changes @ecoussoux-ansys I left some minor comments

ecoussoux-ansys committed:
Update .github/workflows/ci_cd.yml

Co-authored-by: Sébastien Morais <146729917+SMoraisAnsys@users.noreply.github.com>

ecoussoux-ansys committed:
ci: Apply suggestions from code review

SMoraisAnsys commented:
@ecoussoux-ansys Same comment as in https://github.com/ansys/pyaedt-toolkits-common/pull/362#issuecomment-3626998703

SMoraisAnsys commented on .github/workflows/ci_cd.yml:
```diff
@@ -49,7 +63,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
-      - name: Exit if dependabot triggered the workflow
+      - name: Exit if dependabot triggered the workflow # zizmor: ignore[bot-conditions] safe ignore because we exit if dependabot is the triggering action
```
```suggestion
      - name: Exit if dependabot triggered the workflow # zizmor: ignore[bot-conditions] safe ignore because we exit if dependabot is the triggering actor
```

SMoraisAnsys committed:
Apply suggestion from @SMoraisAnsys

ecoussoux-ansys committed:
ci: Restore env variable mistakenly removed in linux tests

ecoussoux-ansys commented:
> @ecoussoux-ansys Same comment as in [ansys/pyaedt-toolkits-common#362 (comment)](https://github.com/ansys/pyaedt-toolkits-common/pull/362#issuecomment-3626998703)

Thanks much for catching this here too, I restored it in a40e9db.

ecoussoux-ansys committed:
ci: Update actions and make check-actions-security dependent on dependabot checks

ecoussoux-ansys commented:
As mentioned in https://github.com/ansys/pyaedt-toolkits-antenna/pull/356#discussion_r2614923865, I updated the ``ansys/actions`` to v10.2.3 and additionally also updated the ``actions/checkout`` to v6.0.1 and the ``codecov/codecov-action`` to v5.5.2 in 3e46246.

As mentioned in https://github.com/ansys/pyaedt-toolkits-antenna/pull/356#discussion_r2614930549, I adopted in 3e46246 the same approach as introduced in https://github.com/ansys/pyaedt-toolkits-common/pull/313 for the vulnerability check: The ``actions-security`` job will only be executed if the ``block-dependabot`` and the ``block-pyansys-ci-bot`` checks pass.

ecoussoux-ansys committed:
ci: Fix code style error

gmalinve commented:
closing as duplicate of #297
